modules/ingester/flush.go

```diff
@@ -52,28 +60,52 @@ func (i *Ingester) Flush() {
 	}
 }
 
-// FlushHandler calls sweepUsers(true) which will force push all traces into the WAL and force
+// ShutdownHandler handles a graceful shutdown for an ingester. It does the following things in order
+// * Stop incoming writes by exiting from the ring
+// * Flush all blocks to backend
+func (i *Ingester) ShutdownHandler(w http.ResponseWriter, _ *http.Request) {
+	// lifecycler should exit the ring on shutdown
+	i.lifecycler.SetUnregisterOnShutdown(true)
+
+	// stop accepting new writes
+	i.markUnavailable()
+
+	// move all data into flushQueue
+	i.sweepAllInstances(true)
+
+	for !i.flushQueues.IsEmpty() {
```

Comment:
i think we want to return immediately from the shutdownhandler. 

```
	if i.flushQueues != nil {
		i.flushQueues.Stop()
		i.flushQueuesDone.Wait()
	}
```
does this logic in stopping already do what we want?  prevent the ingester from shutting down while there is work in the queues?

Reply:
> i think we want to return immediately from the shutdownhandler.

ok. will make that change. 

> does this logic in stopping already do what we want? prevent the ingester from shutting down while there is work in the queues?

I think this doesn't work for us because it stops further "enqueues". So if a complete operation is going on, the following flush operation will never be enqueued..